File: min_cycle_ratio_ai.cs
using System.Collections.Generic;
using System.Linq;
using System;

public class MinCycleRatioSolver<V, R>
    where R : struct, IComparable<R>, IEquatable<R>, IConvertible {
    private readonly Dictionary<V, Dictionary<V, dynamic>> gra;
    public MinCycleRatioSolver(Dictionary<V, Dictionary<V, dynamic>> gra) {
        this.gra = gra;
    }
    public (R, List<(V, V)>) Run(Dictionary<V, R> dist, R r0) {
        var omega = new CycleRatioAPI<V, R>(gra, r0.GetType());
        var solver = new MaxParametricSolver<V, R>(gra, omega);
        var (ratio, cycle) = solver.Run(dist, r0);
        return (ratio, cycle);
    }
}

public class CycleRatioAPI<V, R>
    where R : struct, IComparable<R>, IEquatable<R>, IConvertible {
    private readonly Dictionary<V, Dictionary<V, dynamic>> gra;
    private readonly Type T;
    public CycleRatioAPI(Dictionary<V, Dictionary<V, dynamic>> gra, Type T) {
        this.gra = gra;
        this.T = T;
    }
    public R Distance(R ratio, (V, V)e) {
        var (u, v) = e;
        return (R)(dynamic)(gra[u][v]["cost"] - ratio * gra[u][v]["time"]);
    }
    public R ZeroCancel(List<(V, V)> cycle) {
        var totalCost = cycle.Sum(e => gra[e.Item1][e.Item2]["cost"]);
        var totalTime = cycle.Sum(e => gra[e.Item1][e.Item2]["time"]);
        return (R)(dynamic)(Convert.ChangeType(totalCost, T) / totalTime);
    }
}

public static class Extensions {
    public static void
    SetDefault<V, D>(this Dictionary<V, Dictionary<V, dynamic>> gra,
                     string weight, D value)
        where D : struct, IComparable<D>, IEquatable<D>, IConvertible {
        foreach (var u in gra.Keys.ToList()) {
            foreach (var v in gra[u].Keys.ToList()) {
                if (!gra[u][v].ContainsKey(weight)) {
                    gra[u][v][weight] = value;
                }
            }
        }
    }
}

File: min_cycle_ratio_ai.cpp
#include <algorithm>
#include <string>
#include <type_traits>
#include <unordered_map>
#include <vector>

template <typename V, typename R> class CycleRatioAPI {
  private:
    std::unordered_map<V, std::unordered_map<V, std::unordered_map<std::string, R>>> gra;
    std::type_info const &T;

  public:
    CycleRatioAPI(
        std::unordered_map<V, std::unordered_map<V, std::unordered_map<std::string, R>>> gra,
        std::type_info const &T)
        : gra(gra), T(T) {}
    R Distance(R ratio, std::pair<V, V> e) {
        auto [u, v] = e;
        return gra[u][v]["cost"] - ratio * gra[u][v]["time"];
    }
    R ZeroCancel(std::vector<std::pair<V, V>> cycle) {
        R totalCost = 0, totalTime = 0;
        for (auto e : cycle) {
            totalCost += gra[e.first][e.second]["cost"];
            totalTime += gra[e.first][e.second]["time"];
        }
        return static_cast<R>(static_cast<typename std::underlying_type<R>::type>(totalCost)
                              / totalTime);
    }
};

template <typename V, typename R> class MaxParametricSolver {
  private:
    std::unordered_map<V, std::unordered_map<V, std::unordered_map<std::string, R>>> gra;
    CycleRatioAPI<V, R> omega;

  public:
    MaxParametricSolver(
        std::unordered_map<V, std::unordered_map<V, std::unordered_map<std::string, R>>> gra,
        CycleRatioAPI<V, R> omega)
        : gra(gra), omega(omega) {}
    std::pair<R, std::vector<std::pair<V, V>>> Run(std::unordered_map<V, R> dist, R r0) {
        std::vector<std::pair<V, V>> cycle;
        R ratio = r0;
        for (int i = 0; i < 100; i++) {
            std::unordered_map<V, std::unordered_map<V, R>> g;
            for (auto [u, adj] : gra) {
                for (auto [v, e] : adj) {
                    g[u][v] = omega.Distance(ratio, {u, v});
                }
            }
            auto [d, p] = dist;
            auto [d2, p2] = dist;
            for (int i = 0; i < gra.size(); i++) {
                for (auto [u, adj] : gra) {
                    for (auto [v, e] : adj) {
                        if (d2[v] > d[u] + g[u][v]) {
                            d2[v] = d[u] + g[u][v];
                            p2[v] = u;
                        }
                    }
                }
                std::swap(d, d2);
                std::swap(p, p2);
            }
            V x = V();
            for (auto [u, adj] : gra) {
                for (auto [v, e] : adj) {
                    if (d[u] + g[u][v] < d[v]) {
                        x = u;
                        break;
                    }
                }
                if (x != V()) break;
            }
            if (x == V()) break;
            std::vector<V> path;
            std::unordered_map<V, int> vis;
            while (!vis.count(x)) {
                vis[x] = path.size();
                path.push_back(x);
                x = p[x];
            }
            cycle.clear();
            for (int i = vis[x]; i < path.size(); i++) {
                cycle.push_back({path[i], p[path[i]]});
            }
            R r = omega.ZeroCancel(cycle);
            if (r > ratio) {
                ratio = r;
            } else {
                break;
            }
        }
        return {ratio, cycle};
    }
};

template <typename V, typename R> class MinCycleRatioSolver {
  private:
    std::unordered_map<V, std::unordered_map<V, std::unordered_map<std::string, R>>> gra;

  public:
    MinCycleRatioSolver(
        std::unordered_map<V, std::unordered_map<V, std::unordered_map<std::string, R>>> gra)
        : gra(gra) {}
    std::pair<R, std::vector<std::pair<V, V>>> Run(std::unordered_map<V, R> dist, R r0) {
        CycleRatioAPI<V, R> omega(gra, typeid(r0));
        MaxParametricSolver<V, R> solver(gra, omega);
        auto [ratio, cycle] = solver.Run(dist, r0);
        return {ratio, cycle};
    }
};
